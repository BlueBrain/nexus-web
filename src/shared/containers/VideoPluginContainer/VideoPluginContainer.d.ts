import * as React from 'react';
import { Resource } from '@bbp/nexus-sdk';
import '../../styles/video-plugin.less';
declare type VideoProps = {
  orgLabel: string;
  projectLabel: string;
  resource: Resource;
  collapsed: boolean;
  handleCollapseChanged: () => void;
};
declare const VideoPluginContainer: React.FunctionComponent<VideoProps>;
export default VideoPluginContainer;
