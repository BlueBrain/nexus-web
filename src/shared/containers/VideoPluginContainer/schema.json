{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://bbp.epfl.ch/nexus/plugins/video/schema/video.schema.json",
  "type": "object",
  "properties": {
    "video": {
      "oneOf": [
        { "$ref": "#/$defs/video" },
        { "type": "array", "items": { "$ref": "#/$defs/video" } }
      ]
    }
  },
  "required": ["video"],

  "$defs": {
    "video": {
      "type": "object",
      "properties": {
        "name": {
          "type": "string"
        },
        "description": {
          "type": "string"
        },
        "duration": {
          "type": "string"
        },
        "uploadDate": {
          "type": "string",
          "format": "date-time"
        },
        "embedUrl": {
          "type": "string"
        }
      },
      "required": ["name", "embedUrl"]
    }
  }
}
