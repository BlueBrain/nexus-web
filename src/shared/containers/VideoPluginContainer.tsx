import * as React from 'react';
import { Resource } from '@bbp/nexus-sdk';
import { useNexusContext } from '@bbp/react-nexus';
import ReactPlayer from 'react-player';
import * as moment from 'moment';
import { Collapse, Modal, Button, List, Avatar } from 'antd';

import '../styles/video-plugin.less';

const { Panel } = Collapse;

type VideoProps = {
  orgLabel: string;
  projectLabel: string;
  resource: Resource;
};

type VideoObject = {
  name: String;
  description: String;
  thumbnailUrl: [];
  uploadDate: Date;
  duration: String;
  embedUrl: String;
};

const VideoPluginContainer: React.FunctionComponent<VideoProps> = ({
  resource,
  orgLabel,
  projectLabel,
}) => {
  const nexus = useNexusContext();
  const [videoData, setVideoData] = React.useState<any>();
  const [isModalVisible, setIsModalVisible] = React.useState<boolean>(false);
  const [selectedVideo, setSelectedVideo] = React.useState<any>();

  const handleOk = () => {
    const isModalVisible = false;
    setIsModalVisible(isModalVisible);
  };

  const handleCancel = () => {
    setIsModalVisible(false);
    setSelectedVideo(null);
  };

  React.useEffect(() => {
    loadVideo();
  }, []);

  const handleSelectedVideo = (video: any) => {
    const selectedVideo = video;
    setSelectedVideo(selectedVideo);
    setIsModalVisible(true);
  };
  const loadVideo = async () => {
    const videoResource = (await nexus.Resource.get(
      orgLabel,
      projectLabel,
      encodeURIComponent(resource['@id'])
    )) as Resource<{
      video: VideoObject[];
    }>;
    const videoData = videoResource['video']
      ? (videoResource['video'] as VideoObject[])
      : undefined;
    setVideoData(videoData);
  };

  if (!videoData) return null;
  return (
    <Collapse onChange={() => {}}>
      {videoData && videoData[0]['embedUrl'] ? (
        <Panel header="Video" key="1">
          <List
            itemLayout="horizontal"
            dataSource={videoData}
            renderItem={(item: any) => (
              <List.Item
                extra={
                  <div>
                    <p>{moment.duration(item.duration).humanize()}</p>
                    <p>{moment(item.uploadDate).format('DD/MM/YYYY')}</p>
                  </div>
                }
              >
                <List.Item.Meta
                  avatar={
                    <Avatar
                      style={{
                        width: '100%',
                      }}
                      src={item.thumbnailUrl[0]}
                      shape="square"
                      size={100}
                    />
                  }
                  title={
                    <Button
                      type="link"
                      onClick={() => {
                        handleSelectedVideo(item);
                      }}
                    >
                      {item.name}
                    </Button>
                  }
                  description={item.description}
                />
              </List.Item>
            )}
          />
          {selectedVideo && !!selectedVideo['name'] ? (
            <Modal
              title={selectedVideo['name']}
              bodyStyle={{ padding: 0 }}
              visible={isModalVisible && !!selectedVideo}
              onOk={handleOk}
              onCancel={handleCancel}
              width={640}
              footer={null}
            >
              {!!selectedVideo['embedUrl'] ? (
                <ReactPlayer url={selectedVideo['embedUrl']} />
              ) : null}
            </Modal>
          ) : null}
        </Panel>
      ) : null}
    </Collapse>
  );
};

export default VideoPluginContainer;
