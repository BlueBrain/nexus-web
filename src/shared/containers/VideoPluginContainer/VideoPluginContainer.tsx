import * as React from 'react';
import { Resource } from '@bbp/nexus-sdk';
import { useNexusContext } from '@bbp/react-nexus';
import ReactPlayer from 'react-player';
import * as moment from 'moment';
import { Collapse, Modal, Button, List, notification } from 'antd';
import '../../styles/video-plugin.less';
import { getDateString } from '../../utils';
import * as schema from './schema.json';
import SchemaValidationFallbackContainer from '../SchemaValidationFallbackContainer';
const { Panel } = Collapse;

type VideoProps = {
  orgLabel: string;
  projectLabel: string;
  resource: Resource;
  collapsed: boolean;
  handleCollapseChanged: () => void;
};

type VideoObject = {
  name: string;
  description: string;
  thumbnailUrl: [];
  uploadDate: Date;
  duration: string;
  embedUrl: string;
};

const VideoPluginContainer: React.FunctionComponent<VideoProps> = ({
  resource,
  orgLabel,
  projectLabel,
  collapsed,
  handleCollapseChanged,
}) => {
  const nexus = useNexusContext();
  const [videoData, setVideoData] = React.useState<any>();
  const [isModalVisible, setIsModalVisible] = React.useState<boolean>(false);
  const [selectedVideo, setSelectedVideo] = React.useState<any>();

  const handleOk = () => {
    const isModalVisible = false;
    setIsModalVisible(isModalVisible);
  };

  const handleCancel = () => {
    setIsModalVisible(false);
    setSelectedVideo(null);
  };

  React.useEffect(() => {
    loadVideo();
  }, []);

  const handleSelectedVideo = (video: any) => {
    const selectedVideo = video;
    setSelectedVideo(selectedVideo);
    setIsModalVisible(true);
  };
  const loadVideo = async () => {
    try {
      const videoResource = (await nexus.Resource.get(
        orgLabel,
        projectLabel,
        encodeURIComponent(resource['@id'])
      )) as Resource<{
        video: VideoObject[];
      }>;

      const videoData = videoResource.video
        ? ([videoResource.video].flat() as VideoObject[])
        : undefined;

      setVideoData(videoData);
    } catch (e) {
      notification.info({
        message: 'Error fetching video data',
      });
    }
  };

  return (
    <Collapse
      activeKey={collapsed ? 'video' : undefined}
      onChange={handleCollapseChanged}
    >
      <Panel header="Video" key="video" forceRender={true}>
        <SchemaValidationFallbackContainer
          schema={schema}
          resource={resource}
          dependencies={[resource, projectLabel, orgLabel]}
        >
          <>
            <List
              itemLayout="horizontal"
              dataSource={videoData}
              renderItem={(item: any) => {
                return (
                  <List.Item
                    extra={
                      item.duration &&
                      item.uploadDate && (
                        <div>
                          <p>{moment.duration(item.duration).humanize()}</p>
                          <p>
                            {getDateString(item.uploadDate, { noTime: true })}
                          </p>
                        </div>
                      )
                    }
                  >
                    <List.Item.Meta
                      avatar={<ReactPlayer url={item.embedUrl} light={true} />}
                      title={
                        <Button
                          type="link"
                          onClick={() => {
                            handleSelectedVideo(item);
                          }}
                        >
                          {item.name ? item.name : 'Video Name'}
                        </Button>
                      }
                      description={
                        item.description
                          ? item.description
                          : 'Description of video when information available'
                      }
                    />
                  </List.Item>
                );
              }}
            />
            {selectedVideo && !!selectedVideo.name ? (
              <Modal
                title={selectedVideo.name}
                bodyStyle={{ padding: 0 }}
                visible={isModalVisible && !!selectedVideo}
                onOk={handleOk}
                onCancel={handleCancel}
                width={640}
                footer={null}
              >
                {!!selectedVideo.embedUrl ? (
                  <ReactPlayer url={selectedVideo.embedUrl} />
                ) : null}
              </Modal>
            ) : null}
          </>
        </SchemaValidationFallbackContainer>
      </Panel>
    </Collapse>
  );
};

export default VideoPluginContainer;
