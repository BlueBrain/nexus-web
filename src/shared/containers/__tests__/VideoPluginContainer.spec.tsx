import * as React from 'react';
import { NexusProvider } from '@bbp/react-nexus';
import { createNexusClient, Resource } from '@bbp/nexus-sdk';
import VideoPluginContainer from '../VideoPluginContainer/VideoPluginContainer';
import { QueryClient, QueryClientProvider } from 'react-query';
import fetch from 'node-fetch';
import {
  render,
  fireEvent,
  waitFor,
  screen,
  server,
} from '../../../utils/testUtil';
import { rest } from 'msw';
import '@testing-library/jest-dom';
import { act } from 'react-dom/test-utils';
describe('VideoPluginContainer', () => {
  // establish API mocking before all tests
  beforeAll(() => server.listen());
  // reset any request handlers that are declared as a part of our tests
  // (i.e. for testing one-time error scenarios)
  afterEach(() => server.resetHandlers());
  // clean up once the tests are done
  afterAll(() => server.close());

  const nexus = createNexusClient({
    fetch,
    uri: 'https://localhost:3000',
  });

  it('renders with well formatted data', async () => {
    const resource = ({
      '@context': [
        'https://bbp.neuroshapes.org',
        'https://bluebrain.github.io/nexus/contexts/resource.json',
      ],
      '@id': 'videoresourceId',
      '@type': [
        'Entity',
        'InVitroSliceReconstructedPatchedNeuron',
        'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/InVitroSliceReconstructedPatchedNeuron',
        'ReconstructedPatchedCell',
        'ReconstructedCell',
        'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/ReconstructedPatchedCell',
        'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/ReconstructedCell',
        'Dataset',
      ],
      video: [
        {
          name: 'video',
          embedUrl: 'https://youtu.be/d4gSor3KyIw',
        },
        {
          name: 'cool brain video',

          embedUrl: 'https://www.example.com/embed/123',
        },
      ],
      distribution: [
        {
          '@type': 'DataDownload',
          contentSize: {
            unitCode: 'bytes',
            value: 245427,
          },
          contentUrl:
            'https://bbp.epfl.ch/neurosciencegraph/data/a91043da-0527-4414-b7e8-3484d426c11f',
          digest: {
            algorithm: 'SHA-256',
            value:
              '02c666b5ba8ba806470f77f52b5b6b105662bc63f8b3f2da9a560af0bbc6fc07',
          },
          encodingFormat: 'application/swc',
          name: 'file',
        },
        {
          '@type': 'DataDownload',
          contentUrl: 'http://microcircuits.epfl.ch/#/article/article_3_mph',
          repository: {
            '@id': 'http://microcircuits.epfl.ch/#/article/article_3_mph',
          },
        },
      ],
    } as unknown) as Resource;
    server.use(
      rest.get(
        'https://localhost:3000/resources/org/project/_/videoresourceId',
        (req, res, ctx) => {
          const mockResponse = {
            '@context': [
              'https://bbp.neuroshapes.org',
              'https://bluebrain.github.io/nexus/contexts/resource.json',
            ],
            '@id': 'video_resource_id',
            '@type': [
              'Entity',
              'InVitroSliceReconstructedPatchedNeuron',
              'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/InVitroSliceReconstructedPatchedNeuron',
              'ReconstructedPatchedCell',
              'ReconstructedCell',
              'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/ReconstructedPatchedCell',
              'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/ReconstructedCell',
              'Dataset',
            ],
            video: [
              {
                name: 'video',
                embedUrl: 'https://youtu.be/d4gSor3KyIw',
              },
              {
                name: 'cool brain video',
                embedUrl: 'https://www.example.com/embed/123',
              },
            ],
            distribution: [
              {
                '@type': 'DataDownload',
                contentSize: {
                  unitCode: 'bytes',
                  value: 245427,
                },
                contentUrl:
                  'https://bbp.epfl.ch/neurosciencegraph/data/a91043da-0527-4414-b7e8-3484d426c11f',
                digest: {
                  algorithm: 'SHA-256',
                  value:
                    '02c666b5ba8ba806470f77f52b5b6b105662bc63f8b3f2da9a560af0bbc6fc07',
                },
                encodingFormat: 'application/swc',
                name: 'file',
              },
              {
                '@type': 'DataDownload',
                contentUrl:
                  'http://microcircuits.epfl.ch/#/article/article_3_mph',
                repository: {
                  '@id': 'http://microcircuits.epfl.ch/#/article/article_3_mph',
                },
              },
            ],
          };

          return res(
            // Respond with a 200 status code
            ctx.status(200),
            ctx.json(mockResponse)
          );
        }
      )
    );
    await act(async () => {
      const { container } = await render(
        <NexusProvider nexusClient={nexus}>
          <VideoPluginContainer
            resource={resource}
            orgLabel="org"
            projectLabel="project"
            collapsed={false}
            handleCollapseChanged={jest.fn}
          ></VideoPluginContainer>
        </NexusProvider>
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('fails to render when json does not match schema', async () => {
    const resource = ({
      '@context': [
        'https://bbp.neuroshapes.org',
        'https://bluebrain.github.io/nexus/contexts/resource.json',
      ],
      '@id': 'videoresourceId',
      '@type': [
        'Entity',
        'InVitroSliceReconstructedPatchedNeuron',
        'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/InVitroSliceReconstructedPatchedNeuron',
        'ReconstructedPatchedCell',
        'ReconstructedCell',
        'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/ReconstructedPatchedCell',
        'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/ReconstructedCell',
        'Dataset',
      ],
      video: [
        {
          name: 'video',
          embedurl: ['yyy'],
        },
        {
          name: 'cool brain video',
          embedurl: ['yyy'],
        },
      ],
    } as unknown) as Resource;
    server.use(
      rest.get(
        'https://localhost:3000/resources/org/project/_/videoresourceId',
        (req, res, ctx) => {
          const mockResponse = {
            '@context': [
              'https://bbp.neuroshapes.org',
              'https://bluebrain.github.io/nexus/contexts/resource.json',
            ],
            '@id': 'video_resource_id',
            '@type': ['Entity'],
            video: [
              {
                name: 'video',
                embedurl: ['yyy'],
              },
              {
                name: 'cool brain video',
                embedurl: ['yyy'],
              },
            ],
          };

          return res(
            // Respond with a 200 status code
            ctx.status(200),
            ctx.json(mockResponse)
          );
        }
      )
    );
    await act(async () => {
      const { container } = await render(
        <NexusProvider nexusClient={nexus}>
          <VideoPluginContainer
            resource={resource}
            orgLabel="org"
            projectLabel="project"
            collapsed={false}
            handleCollapseChanged={jest.fn}
          ></VideoPluginContainer>
        </NexusProvider>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
