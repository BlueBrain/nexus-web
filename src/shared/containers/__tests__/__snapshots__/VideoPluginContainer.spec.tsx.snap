// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPluginContainer fails to render when json does not match schema 1`] = `
<div>
  <div
    class="ant-collapse ant-collapse-icon-position-left"
  >
    <div
      class="ant-collapse-item"
    >
      <div
        aria-expanded="false"
        class="ant-collapse-header"
        role="button"
        tabindex="0"
      >
        <span
          aria-label="right"
          class="anticon anticon-right ant-collapse-arrow"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
            />
          </svg>
        </span>
        Video
      </div>
      <div
        class="ant-collapse-content ant-collapse-content-inactive"
        style="display: none;"
      >
        <div
          class="ant-collapse-content-box"
        >
          <div
            class="ant-list ant-list-split"
          >
            <div
              class="ant-spin-nested-loading"
            >
              <div
                class="ant-spin-container"
              >
                <div
                  class="ant-list-empty-text"
                >
                  <div
                    class="ant-empty ant-empty-normal"
                  >
                    <div
                      class="ant-empty-image"
                    >
                      <svg
                        class="ant-empty-img-simple"
                        height="41"
                        viewBox="0 0 64 41"
                        width="64"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <g
                          fill="none"
                          fill-rule="evenodd"
                          transform="translate(0 1)"
                        >
                          <ellipse
                            class="ant-empty-img-simple-ellipse"
                            cx="32"
                            cy="33"
                            rx="32"
                            ry="7"
                          />
                          <g
                            class="ant-empty-img-simple-g"
                            fill-rule="nonzero"
                          >
                            <path
                              d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                            />
                            <path
                              class="ant-empty-img-simple-path"
                              d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                            />
                          </g>
                        </g>
                      </svg>
                    </div>
                    <div
                      class="ant-empty-description"
                    >
                      No Data
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VideoPluginContainer renders with well formatted data 1`] = `
<div>
  <div
    class="ant-collapse ant-collapse-icon-position-left"
  >
    <div
      class="ant-collapse-item"
    >
      <div
        aria-expanded="false"
        class="ant-collapse-header"
        role="button"
        tabindex="0"
      >
        <span
          aria-label="right"
          class="anticon anticon-right ant-collapse-arrow"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
            />
          </svg>
        </span>
        Video
      </div>
      <div
        class="ant-collapse-content ant-collapse-content-inactive"
        style="display: none;"
      >
        <div
          class="ant-collapse-content-box"
        >
          <div
            class="ant-list ant-list-split"
          >
            <div
              class="ant-spin-nested-loading"
            >
              <div
                class="ant-spin-container"
              >
                <div
                  class="ant-list-empty-text"
                >
                  <div
                    class="ant-empty ant-empty-normal"
                  >
                    <div
                      class="ant-empty-image"
                    >
                      <svg
                        class="ant-empty-img-simple"
                        height="41"
                        viewBox="0 0 64 41"
                        width="64"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <g
                          fill="none"
                          fill-rule="evenodd"
                          transform="translate(0 1)"
                        >
                          <ellipse
                            class="ant-empty-img-simple-ellipse"
                            cx="32"
                            cy="33"
                            rx="32"
                            ry="7"
                          />
                          <g
                            class="ant-empty-img-simple-g"
                            fill-rule="nonzero"
                          >
                            <path
                              d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                            />
                            <path
                              class="ant-empty-img-simple-path"
                              d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                            />
                          </g>
                        </g>
                      </svg>
                    </div>
                    <div
                      class="ant-empty-description"
                    >
                      No Data
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
